cluster-autoscaler/core/scaledown/actuation/actuator.go

```diff
@@ -78,160 +79,231 @@ func (a *Actuator) ClearResultsNotNewerThan(t time.Time) {
 
 // StartDeletion triggers a new deletion process.
 func (a *Actuator) StartDeletion(empty, drain []*apiv1.Node) (*status.ScaleDownStatus, errors.AutoscalerError) {
+	emptyIndividual, emptyAtomic := a.groupByNodeGroup(empty)
+	drainIndividual, drainAtomic := a.groupByNodeGroup(drain)
+
 	deletionStartTime := time.Now()
 	defer func() { metrics.UpdateDuration(metrics.ScaleDownNodeDeletion, time.Now().Sub(deletionStartTime)) }()
 
 	results, ts := a.nodeDeletionTracker.DeletionResults()
 	scaleDownStatus := &status.ScaleDownStatus{NodeDeleteResults: results, NodeDeleteResultsAsOf: ts}
 
-	emptyToDelete, drainToDelete := a.cropNodesToBudgets(empty, drain)
-	if len(emptyToDelete) == 0 && len(drainToDelete) == 0 {
+	emptyIndividualToDelete, drainIndividualToDelete, emptyAtomicToDelete, drainAtomicToDelete := a.cropNodesToBudgets(emptyIndividual, drainIndividual, emptyAtomic, drainAtomic)
```

Comment:
@x13n 
My idea of getting rid of the increasing number of pools here would be to create a dedicated struct for a bucket of nodes to scale down, for example:

`type nodeBucket struct{
  nodeGroup cloudprovider.NodeGroup
  nodes []*apiv1.Node
  atomic bool
  drain bool
}`

This could be populated by the Planner and processed by each component in appropriate order (e.g. cropNodesToBudgets would go through atomic first, delete would go through empty first). I think we could also move `cropNodesToBudgets()` out of the actuator.go if we want to keep it's size limited, the rest seems more tightly coupled with the actuation logic. Please let me know WDYT.

Reply:
I moved it out of the main actuator object, however I wanted to limit the amount of changes this PR introduces, so I still left it in the actuator directory. Please LMK if you're OK with it.

Regarding the batching logic, after our offline discussion I ended up with a wrapper over batcher, that will queue nodes for deletion within one pass of scale-down loop, but it will roll them back if any other node fails.